### PR TITLE
Fix #21: Add robust canonical JSON implementation

### LINKED ISSUE #13: Import Canonical JSON tests
**Motivation:** Needed for Canonical JSON implementation work to be done in Milestone 2

### LINKED ISSUE #21: Cover canonical JSON corner cases
See https://github.com/leplatrem/rs-client.rs/issues/4 and https://github.com/leplatrem/rs-client.rs/issues/3

### PR DESCRIPTION
**Pull request by Vishwa-Mozilla**
Summary: Add serialization logic to cover canonical JSON corner cases

Fixes #13 #21

### CODE REVIEW

**Comment by mostlygeek** [comment]
Two things I would like: 

1. Make canonical_json a separate crate and import it into the rust client 
2. Remove the regex parsing.  I'm not a fan of having regex in low level libraries.  I prefer a custom string tokenizer and serializer.  However, this can be done later.  This is mostly to reduce dependencies and increase performance.

File: src/client/signatures/canonical_json.rs [hunk 1]
```diff
@@ -1,5 +1,433 @@
...
+    debug!("returning number without scientifc notation");
+    return format!("{}", number).replace(r"\.?0+$", "");
+}
+
+fn serialize_child(child: &serde_json::Value) -> String {
+
+    let serialized_child_string = serialize(&child);
+
+    if child.is_string() {
+        return "\"".to_owned() + &serialized_child_string + "\"";
```
Comment (ncalexan) [inline src/client/signatures/canonical_json.rs:30]: What happens to quotes in the serialized child?  Are they already quoted?

Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json.rs:30]: no, they are not quoted

File: src/client/signatures/canonical_json/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,13 @@
...
+name = "canonical_json"
+version = "0.1.0"
+authors = ["Vishwa Patel <vpatel@mozilla.com>"]
+edition = "2018"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+log = "0.4.0"
+env_logger = "0.7.1"
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/Cargo.toml:11]: This should be a dev dependency.

File: src/client/signatures/canonical_json/src/lib.rs [hunk 3]
```diff
@@ -0,0 +1,3 @@
+//! Remote-Settings Canonical JSON library to serialize JSON to String
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/lib.rs:1]: Link to the canonical JSON spec, please.  And explain what canonical JSON is.  The Remote-Settings piece is less important -- say that last, since it's just colour for other consumers.

In fact, while looking around, I found https://github.com/zmanian/canonical_json -- what are the pros and cons of this implementation?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 4]
```diff
@@ -0,0 +1,486 @@
...
+        writer.write(&format_number(value, 10.0_f64.powf(21.0), 10.0_f64.powf(-6.0)).into_bytes())?;
+        Ok(())
+    }
+
+    fn write_char_escape<W: ?Sized>(&mut self, writer: &mut W, char_escape: CharEscape) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write(&"\"".to_owned().into_bytes())?;
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:21]: You're converting a `str` into an owned `String`, converting it into bytes, and then taking a reference.  Use `"foo".as_bytes()` instead.

Generally, Rust uses `to_...` to mean a conversion into an owned thing, and `as_...` to mean a zero- (or low-) cost view onto existing data.

Comment (ncalexan) [reply src/client/signatures/canonical_json/src/ser.rs:21]: (Throughout.)

File: src/client/signatures/canonical_json/src/ser.rs [hunk 5]
```diff
@@ -0,0 +1,486 @@
...
+            CharEscape::LineFeed => {
+                writer.write(&"\n".to_owned().into_bytes())?;
+            },
+            CharEscape::Tab => {
+                writer.write(&"\t".to_owned().into_bytes())?;
+            },
+            CharEscape::CarriageReturn => {
+                writer.write(&"\r".to_owned().into_bytes())?;
+            },
+            _ => { println!("Other char escape"); }
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:35]: Print is definitely wrong.  Is this an error?  Something to silently ignore?  Something to log?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 6]
```diff
@@ -0,0 +1,486 @@
...
+                // its a positive exponent
+                formatted_number_string.pop();
+                formatted_number_string.push_str("e+");
+            }
+
+            formatted_number_string.push(curr_char);
+            prev_char = curr_char;
+        }
+
+        return formatted_number_string.replace(r"\.?0+$", "");
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:70]: Why does this not eat legit trailing 0s, like in `1e10000`?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 7]
```diff
@@ -0,0 +1,486 @@
...
+    }
+
+    debug!("returning number without scientifc notation");
+    return format!("{}", number).replace(r"\.?0+$", "");
+}
+
+fn format_unicode_in_string(serialized_string: String) -> String {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:77]: Might as well make the comment be docstring.  Why not just have this take a writer, rather than building and converting strings in the middle?  Folks who want strings can pass in a `mut String::new()`, since it implements `Write`.

File: src/client/signatures/canonical_json/src/ser.rs [hunk 8]
```diff
@@ -0,0 +1,486 @@
...
+    // looking for \u{X} \u{XX}, \u{XXX}, \u{XXXX} to remove the curly braces
+
+    let mut formatted_string = String::new();
+
+    let mut unicode_start_index = 0;
+
+    while unicode_start_index < serialized_string.len() {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:84]: This probably can work, but it's quite awkward, with the indexing, sizes, and slicing.  Did you try writing this as a loop consuming a `peekable()` iterator of characters?  In pseudo-code:
```
let mut iter = string.into_iter().peekable();
while let Some(char) = iter.next() {
  if char == '\\' && iter.peek() == &'{' {
    // Eat at most four characters.
    ...
    match iter.next() {
      Some('}') => [continue],
      None => [bail out],
    }
  }
}
```

File: src/client/signatures/canonical_json/src/ser.rs [hunk 9]
```diff
@@ -0,0 +1,486 @@
...
+///         "a": "a",
+///         "id": "1",
+///         "b": "b"
+///     })); // returns "{"a":"a","b":"b","id":"1"}"; (orders object keys)
+/// 
+///     serialize(&json!(vec!["one", "two", "three"])); // returns "["one","two","three"]"
+/// # }
+/// 
+/// ```
+pub fn serialize(input: &serde_json::Value) -> String {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:143]: `to_canonical_json`.

File: src/client/signatures/canonical_json/src/ser.rs [hunk 10]
```diff
@@ -0,0 +1,486 @@
...
+        },
+        Err(error) => {
+            error!("Error serializing JSON : {}", error);
+            return null_string;
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:164]: Is there any JSON that cannot be represented in the canonical JSON notation?  If so, consider just having a list of strings in canonical JSON form, and then verifying that the roundtrip String -> `json::Value` -> `to_canonical_json_string` preserves the value.  The named functions don't really add much.

File: src/client/signatures/canonical_json/src/ser.rs (hunk 4 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:21]: I see, yes will make this change

File: src/client/signatures/canonical_json/src/ser.rs (hunk 6 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:70]: I removed this pattern since it was not needed, format("{:e}", number) by default removes trailing zeros after decimal point

Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:70]: yes, the pattern will remove legit trailing zeros which is not expected behaviour

File: src/client/signatures/canonical_json/Cargo.toml (hunk 2 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/Cargo.toml:11]: yes, will make this change in the main crate as well

File: src/client/signatures/canonical_json/src/lib.rs (hunk 3 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/lib.rs:1]: it does not support floating-point numbers, cannot convert Unicode characters to Unicode escape sequences

File: src/client/signatures/canonical_json/src/ser.rs (hunk 8 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:84]: Should I instead create a new string and add to it instead of mutating while traversing an iterable - since when we mutate the iterable, it has to still move elements

Comment (ncalexan) [reply src/client/signatures/canonical_json/src/ser.rs:84]: You can't mutate the iterator itself; it's just a sequence.  (It's `mut` 'cuz every call to `.next()` updates its internal state.)

If you take a `Write` instance as a parameter, then that gives you a place to write the characters that you've processed.

File: src/client/signatures/canonical_json/src/lib.rs (hunk 3 above)
Comment (ncalexan) [reply src/client/signatures/canonical_json/src/lib.rs:1]: Ah!  Then it's definitely worth calling out how this compares to that implementation in the README or similar.

Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/lib.rs:1]: yes

File: src/client/signatures/canonical_json/src/ser.rs (hunk 5 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:35]: other types in CharEscape like backspace and formfeed are not recognized by Rust

Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:35]: Backspace, Formfeed char escapes are not recognized by Rust so there is nothing to do for them

Comment (ncalexan) [reply src/client/signatures/canonical_json/src/ser.rs:35]: If there are a small number of them, then mach them explicitly and do nothing.  If there are a large number, then `_` might be best -- just say what you told me in a comment.

**Comment by Vishwa-Mozilla** [comment]
I mark comments on the PR as resolved when I have made the proposed changes

File: src/client/signatures/canonical_json/src/ser.rs (hunk 5 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs]: yes I have fixed it

File: src/client/signatures/canonical_json/src/ser.rs [hunk 11]
```diff
@@ -0,0 +1,486 @@
...
+
+            formatted_number_string.push(curr_char);
+            prev_char = curr_char;
+        }
+
+        return formatted_number_string.replace(r"\.?0+$", "");
+    }
+
+    debug!("returning number without scientifc notation");
+    return format!("{}", number).replace(r"\.?0+$", "");
```
[the remark this answers is not in the document]
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs]: I have actually removed it since format!() removes unnecessary trailing zeros

File: src/client/signatures/canonical_json/src/ser.rs [hunk 12]
```diff
@@ -0,0 +1,597 @@
...
+    {
+        let formatted_string = format!("{}", fragment).escape_default().to_string().replace("\\\"", "\"");
+        format_unicode_in_string(writer, formatted_string)?;
+        Ok(())
+    }
+}
+
+fn format_number<W: ?Sized>(writer: &mut W, number: f64, lower_bound: f64, upper_bound: f64) -> Result<(), std::io::Error> where W: Write, {
+
+    debug!("in format_number function");
```
Comment (leplatrem) [inline src/client/signatures/canonical_json/src/ser.rs:55]: Do you think this will be useful?

File: src/client/signatures/canonical_json/spec.txt [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+        "rust": "❤",
+        "zoo": [
+            "zorilla",
+            "anteater"
+        ]
+    }
+
+  Canonical JSON:
+    "{"abc":9.30258908e-7,\"def\":\"bar\",\"ghi\":1e+21,\"rust\":\"\\u2764\",\"zoo\":[\"zorilla\",\"anteater\"]}"
+4. Grammar
```
Comment (leplatrem) [inline src/client/signatures/canonical_json/spec.txt:35]: nit: Missing blank line

Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/spec.txt:35]: done

File: src/client/signatures/canonical_json/src/ser.rs (hunk 12 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:55]: no, I will remove it

File: src/client/signatures/canonical_json/README.md [hunk 14]
```diff
@@ -0,0 +1,45 @@
...
+Compared to serde_json::to_string
+- String contents are not guaranteed be parsable as UTF-8. Be aware that encoded data may contain escaped unicode characters.
+- Minus Zero is disallowed
+- Object keys must appear in lexiographical order and must not be repeated.
+
+[Link to the Canonical Spec](spec.txt)
+
+Comparsion to other Canonical JSON implementation - https://github.com/zmanian/canonical_json
+
+- This Canonical JSON library supports floating-point numbers, exponents, and can convert Unicode characters into Unicode escape sequences
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/README.md:16]: Which is "This"?  Our implementation I hope?

File: src/client/signatures/canonical_json/README.md [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+- Preserves character escapes for {Tab, CarriageReturn, Newline, LineFeed, Quote}
+- Escapes ReverseSolidus
+- Canonical JSON cannot be parsed by regular JSON parsers
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/README.md:19]: What?  The output of our implementation isn't valid JSON?

File: src/client/signatures/canonical_json/spec.txt [hunk 16]
```diff
@@ -0,0 +1,96 @@
...
+    digit
+    digit1-9 digits
+
+  digits:
+    digit
+    digit digits
+
+5. References
+
+  Spec derived from: http://wiki.laptop.org/go/Canonical_JSON
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/spec.txt:96]: I would just link to this: are we ever going to update this document?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 17]
```diff
@@ -0,0 +1,595 @@
...
+
+    let string = vec![];
+    let null_string = "null".to_owned();
+    let mut serializer = serde_json::Serializer::with_formatter(string, JSONFormatter{});
+    match input.serialize(&mut serializer) {
+        Ok(()) => match String::from_utf8(serializer.into_inner()) {
+            Ok(serialized_string) => return serialized_string,
+            Err(error) => {
+                error!("Error converting bytes to string: {}", error);
+                return null_string;
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:165]: This doesn't make much sense to me: why not return `Result<String, ...>` and let consumers decide if they want this `null_string` value?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 18]
```diff
@@ -0,0 +1,595 @@
...
+    writer.write(&format!("{}", number).into_bytes())?;
+    Ok(())
+}
+
+/// looking for \u{X} \u{XX}, \u{XXX}, \u{XXXX} to remove the curly braces
+fn format_unicode_in_string<W: ?Sized>(writer: &mut W, serialized_string: String) -> Result<(), std::io::Error> where W: Write, {
+
+    let mut string_iter = serialized_string.chars().peekable();
+
+    while let Some(curr_char) = string_iter.next() {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:83]: nit: Tighten the vertical whitespace hre, no blank lanes after loop/block starts, please.

Comment (ncalexan) [reply src/client/signatures/canonical_json/src/ser.rs:83]: Better: run `rustfmt`.

File: src/client/signatures/canonical_json/src/ser.rs [hunk 19]
```diff
@@ -0,0 +1,595 @@
...
+        }
+
+        Ok(())
+    }
+
+    fn write_string_fragment<W: ?Sized>(&mut self, writer: &mut W, fragment: &str) -> Result<(), std::io::Error> where
+        W: Write, 
+    {
+        let formatted_string = format!("{}", fragment).escape_default().to_string().replace("\\\"", "\"");
+        format_unicode_in_string(writer, formatted_string)?;
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:48]: This returns a result already -- try `format_unicode(...).and(())`

File: src/client/signatures/canonical_json/src/ser.rs [hunk 20]
```diff
@@ -0,0 +1,595 @@
...
+}
+
+#[cfg(test)]
+mod tests {
+    use serde_json::json;
+    use log::{debug};
+    use env_logger;
+    use super::{to_canonical_json};
+
+    struct Test<'a> {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:182]: Don't change it now, but for next time, there's still a lot of boilerplate here: a `Test` type (with a lifetime, no less!) and a description.  You've got the same number of lines as the functions.  Next time, drop the descriptions -- if you need, that's what comments are for.

The idiomatic way to do this is to use a Rust macro to be able to declare:

```rust
test!(input1, output1);
test!(input2, output2);
...
```
or similar.  And that's what I want: just 50 lines of pairs.  But now that it's here, don't bother -- you can spend a long time fighting Rust's macro system.

File: src/client/signatures/canonical_json/src/ser.rs [hunk 21]
```diff
@@ -0,0 +1,595 @@
...
+                input: json!(vec![json!({ "key": "é" })]),
+                output: "[{\"key\":\"\\u00e9\"}]"
+            },
+            Test {
+                test_description: "to_canonical_json_serialize_array_preserves_data",
+                input: json!(vec![
+                    json!({ "foo": "bar", "last_modified": "12345", "id": "1" }),
+                    json!({ "bar": "baz", "last_modified": "45678", "id": "2" }),
+                ]),
+                output: "[{\"foo\":\"bar\",\"id\":\"1\",\"last_modified\":\"12345\"},\
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:574]: Use [raw string literals](https://rahul-thakoor.github.io/rust-raw-string-literals/) for anything with quotes, please.

File: src/client/signatures/canonical_json/README.md (hunk 14 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/README.md:16]: yes 😄

File: src/client/signatures/canonical_json/README.md (hunk 15 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/README.md:19]: 😨 I need to replace that "cannot" with "can"

File: src/client/signatures/canonical_json/spec.txt (hunk 16 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/spec.txt:96]: I don't think so but I have added the float type in the grammar which wasn't there in the spec link

File: src/client/signatures/canonical_json/src/ser.rs (hunk 18 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:83]: done

File: src/client/signatures/canonical_json/src/ser.rs (hunk 21 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:574]: yes

**Comment by leplatrem** [comment]
Quick nit:  after having used the canonical_json crate in https://github.com/leplatrem/python-canonicaljson-rs/pull/2 I think we could rename `canonical_json::to_canonical_json()` to `canonical_json::to_string()`

File: README.md [hunk 22]
```diff
@@ -40,11 +40,10 @@ async fn main() {
 
 ### Logging
 
-Dependencies: [log](https://docs.rs/log), [env_logger](https://docs.rs/env_logger)
+Dependencies: [env_logger](https://docs.rs/env_logger)
```
Comment (ncalexan) [inline README.md:43]: This means that `log` is a transitive dependency, just FYI.

File: src/client/signatures/canonical_json/src/ser.rs [hunk 23]
```diff
@@ -0,0 +1,421 @@
...
+    Ok(serialized_string)
+}
+
+#[cfg(test)]
+mod tests {
+    use super::to_string;
+    use env_logger;
+    use serde_json::json;
+
+    macro_rules! test_canonical_json {
```
Comment (ncalexan) [inline src/client/signatures/canonical_json/src/ser.rs:212]: Ah, you did this -- neat.

**Comment by leplatrem** [comment]
Before merging, could you confirm that signatures can be verified on [all existing collections](https://firefox.settings.services.mozilla.com/v1/buckets/monitor/collections/changes/records)) with this new serializer?

When using the canonical_json crate via Python bindings I have a few that are failing...

```
  "main/normandy-recipes": "SignatureDoesNotMatch()",
  "main/normandy-recipes-capabilities": "SignatureDoesNotMatch()",
  "main/messaging-experiments": "SignatureDoesNotMatch()",
  "main/cfr": "SignatureDoesNotMatch()",
  "blocklists/addons": "SignatureDoesNotMatch()",
  "blocklists/plugins": "SignatureDoesNotMatch()",
  "main/whats-new-panel": "SignatureDoesNotMatch()",
  "security-state/intermediates": "SignatureDoesNotMatch()",
  "blocklists/certificates": "SignatureDoesNotMatch()",
  "security-state/onecrl": "SignatureDoesNotMatch()",
  "main/fxmonitor-breaches": "SignatureDoesNotMatch()",
  "main/cfr-fxa": "SignatureDoesNotMatch()",
  "main/fftv-experiments": "SignatureDoesNotMatch()"
```

I investigated and it seems it could come from missing backlashes in escape chars (`\n` vs `\\n`)

File: src/client/signatures/canonical_json/src/ser.rs [hunk 24]
```diff
@@ -0,0 +1,421 @@
...
+            CharEscape::Tab => {
+                writer.write("\t".as_bytes())?;
+            }
+            CharEscape::CarriageReturn => {
+                writer.write("\r".as_bytes())?;
+            }
+            CharEscape::Backspace => {}
+            CharEscape::FormFeed => {}
+            CharEscape::Solidus => {}
+            CharEscape::AsciiControl(_number) => {}
```
Comment (leplatrem) [inline src/client/signatures/canonical_json/src/ser.rs:60]: `\u00{_number}` ?

File: src/client/signatures/canonical_json/src/ser.rs [hunk 25]
```diff
@@ -0,0 +1,421 @@
...
+    {
+        match char_escape {
+            CharEscape::Quote => {
+                writer.write("\"".as_bytes())?;
+            }
+            CharEscape::ReverseSolidus => {
+                writer.write("\\\\".as_bytes())?;
+            }
+            CharEscape::LineFeed => {
+                writer.write("\n".as_bytes())?;
```
Comment (leplatrem) [inline src/client/signatures/canonical_json/src/ser.rs:49]: This should be `\\n`

File: src/client/signatures/canonical_json/src/ser.rs [hunk 26]
```diff
@@ -0,0 +1,421 @@
...
+            }
+            CharEscape::Tab => {
+                writer.write("\t".as_bytes())?;
```
Comment (leplatrem) [inline src/client/signatures/canonical_json/src/ser.rs:52]: This should be `\\t`

File: src/client/signatures/canonical_json/src/ser.rs (hunk 25 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:49]: yes

File: src/client/signatures/canonical_json/src/ser.rs (hunk 24 above)
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:60]: added custom serde's ser implemenation for AsciiControl

**Comment by Vishwa-Mozilla** [comment]
yes, I fixed an issue with single quotes being escaped and now sigantures can be verified on all existing collections. Will update the PR

File: src/client/signatures/canonical_json/src/ser.rs [hunk 27]
```diff
@@ -0,0 +1,421 @@
...
+            }
+            CharEscape::CarriageReturn => {
+                writer.write("\r".as_bytes())?;
+            }
+            CharEscape::Backspace => {}
+            CharEscape::FormFeed => {}
+            CharEscape::Solidus => {}
```
[the remark this answers is not in the document]
Comment (Vishwa-Mozilla) [reply src/client/signatures/canonical_json/src/ser.rs:59]: no, I haven't added tests for these as \b, \f are not recognized by Rust
